calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,380 @@
+///       wait_*;
+///       restore_*;
+///     }
+///    or
+///     seq {
+///       <stmt>;
+///       incr_barrier_*_*;
+///       write_barrier_*_*;
+///       wait_*;
+///       wait_restore_*;
+///     }
+
+pub struct CompileSync;
+
+/// the structure used to store cells and groups shared by one barrier
+type BarrierMap = LinkedHashMap<u64, (Vec<RRC<ir::Cell>>, Vec<RRC<ir::Group>>)>;
+
+impl Named for CompileSync {
+    fn name() -> &'static str {
+        "compile-sync"
+    }
+
+    fn description() -> &'static str {
+        "Implement barriers for statements marked with @sync attribute"
+    }
+}
+
+fn build_barriers(
+    builder: &mut ir::Builder,
+    s: &mut ir::Control,
+    map: &mut BarrierMap,
+    count: &mut HashMap<u64, u64>,
+) {
+    match s {
+        ir::Control::Seq(seq) => {
+            // go through each control statement
+            // if @sync
+            // see if we already have the set of shared primitives and groups
+            // initialized
+            // True -> generate the inidividual groups and buikld the seq stmt
+            // False -> generate the shared groups, cells
+            //          put the shared groups in the barriermap
+            //          generate the individual groups
+            //          build the seq stmt
+            let mut stmts_new: Vec<ir::Control> = Vec::new();
+            for stmt in seq.stmts.drain(..) {
+                if let Some(n) = stmt.get_attributes().unwrap().get("sync") {
+                    if map.get(n).is_none() {
+                        add_shared_structure(builder, n, map);
+                    }
+                    let (cells, groups) = map.get(n).unwrap();
+                    let barrier = Rc::clone(&cells[0]);
+                    let eq = Rc::clone(&cells[1]);
+                    let wait_restore = Rc::clone(&groups[0]);
+                    let restore = Rc::clone(&groups[1]);
+                    let clear_barrier = Rc::clone(&groups[2]);
+                    let n_new = *(count.entry(*n).or_insert(0)) + 1;
+                    count.insert(*n, n_new);
+                    structure!(builder;
+                            let save = prim std_reg(32););
+                    let incr_barrier = builder.add_group("incr_barrier");
+                    build_incr_barrier(
+                        builder,
+                        &incr_barrier,
+                        &barrier,
+                        &save,
+                        &(&n_new - 1),
+                    );
+                    let write_barrier = builder.add_group("write_barrier");
+                    build_write_barrier(
+                        builder,
+                        &write_barrier,
+                        &barrier,
+                        &save,
+                        &(&n_new - 1),
+                    );
+                    let wait = builder.add_group("wait");
+                    build_wait(builder, &wait, &eq);
+                    if n_new == 1 {
```

Comment:
This conditional call to two methods that look almost the same is sketchy. Can we generalize the logic between the two? Is this because the first time we observe a `@sync` we have to do something different than the second time we observe it?